### ISSUE TITLE
AbstractControl maintains validator after corresponding element is removed from view

### ISSUE THREAD

**Issue by marcovdb** [angular/angular#14955]
**I'm submitting a ...**  (check one with "x")
```
[*] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
I have a reactive form where parts of the form are conditionally hidden or shown. I have noticed that when an element has a `required` attribute, the 'required' validator is automatically added to its corresponding control. But when the element is removed from the DOM, the form control retains its required validator, and the form will forever remain invalid.

**Expected behavior**
Since the validator gets added automatically to the control, I would expect it to get torn down automatically when the element is removed from the DOM as well.

**Minimal reproduction of the problem with instructions**

http://plnkr.co/edit/qHjP5L34qTIZu8ssUS2M?p=preview
Enable "bar", which is a required input. Then disable it again and see how the form is still invalid.

**What is the motivation / use case for changing the behavior?**
I have a dynamic form that is built using a template. The user selects an item type and based on this type, the form is reset using a predefined set of default values for the chosen item type. In the template, form elements are hidden or shown based on the chosen type. If a required element is hidden, it remains required and the user is unable to submit the form.

**Please tell us about your environment:**

* **Angular version:** 2.4.9

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

All browsers.
 
* **Language:** [all | TypeScript X.X | ES6/7 | ES5]
TS 2.0.10

* **Node (for AoT issues):** `node --version` =   
6.9.5

**Comment by k-schneider** [comment]
Any traction on this?  It's a pretty big issue if you have a component that directly uses NG_VALIDATORS inside of an ngIf.  Every time the component is created the control adds one validator.

Are there any have decent workarounds for this or is the only option to constantly clear the validators and build them again on any condition that causes the control to show/hide?

Thanks.

**Comment by rodrigo-vizbelis** [comment]
There is a similar issue with the DefaultValueAccessor registerOnChange() method. On every init new listener will be added to the FormControl _onChange list and the writeValue() method will be executed multiple times.

**Comment by mtraynham** [comment]
Just encountered this in a different scenario.  I was following along with this blog on how to implement nested forms, a way to split up complex forms with ReactiveForms.
https://blog.angularindepth.com/angular-nested-reactive-forms-using-cvas-b394ba2e5d0d

There's a section in there on validation, which suggests the nested form component implement Validator to pass validation back up to the parent form, e.g. using:

```ts
        providers: [{
            provide: NG_VALIDATORS,
            useExisting: forwardRef(() => Type),
            multi: true
        }]
```

When using these nested forms with `*ngIf` or `[ngSwitch]`, which may remove the nested form control or reuse the control with a different nested form, the validators build up overtime on the parent form as they are never removed from these controls.

Because the parent form control is hanging on to the validator of the nested form, it's actually hanging on to the whole nested form component instance.  This actually seems like a pretty bad memory leak given the right circumstances.

**Comment by VyTacka** [comment]
> Just encountered this in a different scenario. I was following along with this blog on how to implement nested forms, a way to split up complex forms with ReactiveForms.
> https://blog.angularindepth.com/angular-nested-reactive-forms-using-cvas-b394ba2e5d0d
> 
> There's a section in there on validation, which suggests the nested form component implement Validator to pass validation back up to the parent form, e.g. using:
> 
> ```ts
>         providers: [{
>             provide: NG_VALIDATORS,
>             useExisting: forwardRef(() => Type),
>             multi: true
>         }]
> ```
> 
> When using these nested forms with `*ngIf` or `[ngSwitch]`, which may remove the nested form control or reuse the control with a different nested form, the validators build up overtime on the parent form as they are never removed from these controls.
> 
> Because the parent form control is hanging on to the validator of the nested form, it's actually hanging on to the whole nested form component instance. This actually seems like a pretty bad memory leak given the right circumstances.
> 
> I believe the validators are registered when using the following directives `formControl`, `formControlName`, `formGroup`, `formGroupName`, which composes the validators but never decomposes them like [here](https://github.com/angular/angular/blob/master/packages/forms/src/directives/reactive_directives/form_control_directive.ts#L212).

To workaround this issue you can inject `NgControl` into your component which extends `ControlValueAccessor` and then clear `NgControl` validators when component is destroyed. 

Example:
```ts
export class CustomControlComponent implements ControlValueAccessor, Validator, OnInit, OnDestroy {
  private ngControl: NgControl;

  constructor(private injector: Injector) { }

  ngOnInit(): void {
    this.ngControl = this.injector.get(NgControl);
  }

  ngOnDestroy(): void {
    this.ngControl.control.clearValidators();
  }
}
```

Just beware that this can probably clear validators that you have added to this control from parent form component.

**Comment by maxime1992** [comment]
I came across something very similar recently... https://github.com/angular/angular/issues/20007#issuecomment-683496790

This issue could be related I assume.

**Comment by AndrewKushnir** [comment]
This issue should be resolved by PR #39235 which was merged recently and the fix will be available in upcoming v11.1.0 release later this month.
